src/cli/Peers.ts

```diff
@@ -0,0 +1,387 @@
+          throw Error(`agent status is: ${status}`);
+        }
+
+        const nodePath = determineNodePath(options);
+        const publicKeyPath = options.publicKey;
+        // read in publicKey if it exists
+        const publicKey = fs.readFileSync(publicKeyPath).toString();
+
+        const successful = await client.findPeer(nodePath, publicKey);
+
+        if (successful) {
+          pkLogger('peer successfully pinged', PKMessageType.SUCCESS);
+        } else {
+          pkLogger('ping timed out', PKMessageType.WARNING);
+        }
+      }),
+    );
+}
+
+function makeFindSocialPeerCommand() {
+  return new commander.Command('social')
+    .description('find a peer based on a handle and service, e.g. john.smith and github')
+    .option('--node-path <nodePath>', 'node path')
+    .requiredOption('-h, --handle <handle>', 'handle of the user on the specified service, e.g. john.smith')
+    .requiredOption('-s, --service <service>', 'service where the handle can be found, e.g. github')
+    .option('-v, --verbose', 'increase verbosity level by one')
+    .action(
+      actionRunner(async (options) => {
+        const client = PolykeyAgent.connectToAgent();
+        const status = await client.getAgentStatus();
+        if (status != 'online') {
+          throw Error(`agent status is: ${status}`);
+        }
+
+        const nodePath = determineNodePath(options);
+        const handle = options.handle;
+        const service = options.service;
+
+        const successful = await client.findSocialPeer(nodePath, handle, service);
+
+        if (successful) {
+          pkLogger('peer successfully pinged', PKMessageType.SUCCESS);
+        } else {
+          pkLogger('ping timed out', PKMessageType.WARNING);
+        }
+      }),
+    );
+}
+
+function makeStealthCommand() {
+  return new commander.Command('stealth')
+    .description('toggle stealth mode on or off')
+    .option('--node-path <nodePath>', 'node path')
+    .option('--active', 'toggle stealth mode to active')
+    .option('--no-active', 'toggle stealth mode to not active')
+    .option('-v, --verbose', 'increase verbosity level by one')
+    .action(
+      actionRunner(async (options) => {
+        const client = PolykeyAgent.connectToAgent();
+        const status = await client.getAgentStatus();
+        if (status != 'online') {
+          throw Error(`agent status is: ${status}`);
+        }
+
+        const nodePath = determineNodePath(options);
+        const active = options.active;
+        if (active == undefined) {
+          throw Error('active status must be specified with either --active or --no-active');
+        }
+
+        const successful = await client.toggleStealth(nodePath, active);
+
+        pkLogger(`stealth mode toggled to '${active ? 'active' : 'inactive'}'`, PKMessageType.SUCCESS);
+      }),
+    );
+}
+
+function makeRelayCommand() {
+  return new commander.Command('relay')
+    .description('request a relay connection from a public peer')
```

Comment:
I'm thinking NAT based stuff should be as automatic and as invisibile as possible. I don't really want the user to be thinking about NAT.

On the other hand, these things are going to have to be configurable, and they may form as part of a default bootstrap NAT relay servers, but then there's extra configuration allowed to add new relay peers that are available.

Eventually the user should be running their own relay peers. My main idea is that most normal people will be using 2 pk. 1 on mobile, 1 on laptop. Both are likely to need a relay. So I guess most normal people won't be using a relay.

But power users may self-host on the cloud.

Also I want to support NAT-less transfer. I think I see a lot of mobile apps using QR codes. Can we combine QR code visual transfer with some sort of UDP hole punching to get information transfer working without a third party relay?

Reply:
this will be removed, issue is here: https://github.com/MatrixAI/js-polykey/issues/103